src/cdk/table/table.ts

```diff
@@ -162,6 +177,19 @@ export class CdkTable<T> implements CollectionViewer, OnInit, AfterContentChecke
    */
   private _footerRowDefChanged = false;
 
+  /**
+   * Cache of the latest rendered `DataRow` objects as a map for easy retrieval when constructing
+   * a new list of `DataRow` objects for rendering rows. Since the new list is constructed with
+   * the cached `DataRow` objects when possible, the row identity is preserved when the data
+   * and row template matches, which allows the `IterableDiffer` to check rows by reference
+   * and understand which rows are added/moved/removed.
+   *
+   * Implemented as a map of maps where the first key is the `data: T` object and the second is the
+   * `CdkRowDef<T>` object. With the two keys, the cache points to a `DataRow<T>` object that
+   * contains the pair.
+   */
+  private _cachedDataRowsMap = new Map<T, Map<CdkRowDef<T>, DataRow<T>>>();
```

Comment:
Each `T` may map into multiple `DataRow<T>` objects since each object represents a pair of data and row template.

This setup allows us to uniquely identify these pairs so that they can be re-used if they show up again between renders. This is necessary for the `IterableDiffer` to understand the correct operations, e.g. when a pairing is moved.